util/src/main/java/org/killbill/billing/util/discovery/provider/KillbillEurekaInstanceInfoProvider.java

```diff
@@ -0,0 +1,128 @@
+ * Copyright 2014-2018 The Billing Project, LLC
+ *
+ * The Billing Project licenses this file to you under the Apache License, version 2.0
+ * (the "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at:
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.killbill.billing.util.discovery.provider;
+
+import java.util.Map;
+
+import javax.inject.Singleton;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.inject.Inject;
+import com.netflix.appinfo.DataCenterInfo;
+import com.netflix.appinfo.EurekaInstanceConfig;
+import com.netflix.appinfo.InstanceInfo;
+import com.netflix.appinfo.InstanceInfo.InstanceStatus;
+import com.netflix.appinfo.InstanceInfo.PortType;
+import com.netflix.appinfo.LeaseInfo;
+import com.netflix.appinfo.UniqueIdentifier;
+import com.netflix.appinfo.providers.Archaius1VipAddressResolver;
+import com.netflix.appinfo.providers.EurekaConfigBasedInstanceInfoProvider;
+import com.netflix.appinfo.providers.VipAddressResolver;
+
+@Singleton
+public class KillbillEurekaInstanceInfoProvider extends EurekaConfigBasedInstanceInfoProvider {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KillbillEurekaInstanceInfoProvider.class);
+
+    private final EurekaInstanceConfig config;
+
+    private InstanceInfo instanceInfo;
+
+    @Inject(optional = true)
+    private VipAddressResolver vipAddressResolver = null;
+
+    @Inject
+    public KillbillEurekaInstanceInfoProvider(EurekaInstanceConfig config) {
+        super(config);
+        this.config = config;
+    }
+
+    @Override
+    public synchronized InstanceInfo get() {
+        if (instanceInfo == null) {
+            // Build the lease information to be passed to the server based on config
+            LeaseInfo.Builder leaseInfoBuilder = LeaseInfo.Builder.newBuilder()
+                                                                  .setRenewalIntervalInSecs(config.getLeaseRenewalIntervalInSeconds())
+                                                                  .setDurationInSecs(config.getLeaseExpirationDurationInSeconds());
+
+            if (vipAddressResolver == null) {
+                vipAddressResolver = new Archaius1VipAddressResolver();
+            }
+
+            // Builder the instance information to be registered with eureka server
+            InstanceInfo.Builder builder = InstanceInfo.Builder.newBuilder(vipAddressResolver);
+
+            // set the appropriate id for the InstanceInfo, falling back to datacenter Id if applicable, else hostname
+            String instanceId = config.getInstanceId();
```

Comment:
I think we should use the [CreatorName](https://github.com/killbill/killbill-commons/blob/a2f5ea6effc2fa256fe8ace9fb968252c2caf00e/queue/src/main/java/org/killbill/CreatorName.java#L35) instead. This is how instances identifiers are defined today when deploying in multiple datacenters or using AWS Auto Scaling.

FWIW, we also have a `region` property (https://github.com/killbill/killbill-platform/blob/7039977db0b9618a09454673b983c699bbc16740/server/src/main/java/org/killbill/billing/server/config/KillbillServerConfig.java#L44-L47).